### PR TITLE
systemd: fixes for Travis

### PR DESCRIPTION
**Pull request by maxim-belkin**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew-core/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally prior to submission with `brew install <formula>` (where `<formula>` is the name of the formula you're submitting)?
- [x] Does your submission pass `brew audit --strict --online <formula>` (after doing `brew install <formula>`)?

-----

Formulas that depend on `systemd` (e.g. mesa) will fail to pass tests on Travis because of XML::Parser module.

### CODE REVIEW

File: Formula/libusb.rb [hunk 1]
```diff
@@ -29,6 +29,7 @@ class Libusb < Formula
   deprecated_option "no-runtime-logging" => "without-runtime-logging"
 
   depends_on "systemd" if OS.linux? # for libudev
+  depends_on "zlib"    if OS.linux?
```
Comment (sjackman) [inline Formula/libusb.rb]: Please don't align `if`. Collapse multiple spaces to a single space.

Comment (sjackman) [inline Formula/libusb.rb]: Use `unless OS.mac?` rather than `if OS.linux?` for dependencies not required on Mac.

File: Formula/systemd.rb [hunk 2]
```diff
@@ -18,8 +18,22 @@ class Systemd < Formula
   depends_on "util-linux" # for libmount
   depends_on "coreutils" => :build
   depends_on "XML::Parser" => :perl
+  depends_on "expat" => :build unless OS.mac?
+
+  resource "xml::parser" do
+    url "http://search.cpan.org/CPAN/authors/id/M/MS/MSERGEANT/XML-Parser-2.36.tar.gz"
+    sha256 "9fd529867402456bd826fe0e5588d35b3a2e27e586a2fd838d1352b71c2ed73f"
+  end
```
Comment (sjackman) [inline Formula/systemd.rb]: I'd rather `apt-get install libxml-parser-perl` on Travis, if that Perl package is already installed on Mac.

File: Formula/systemd.rb [hunk 3]
```diff
@@ -5,10 +5,6 @@ class Systemd < Formula
   sha256 "085e088650afbfc688ccb13459aedb1fbc7c8810358605b076301f472d51cc4f"
   # tag "linuxbrew"
 
-  bottle do
-    sha256 "b24ff69678f718005768a023d1c264c2486cc69027be407107341c5f648d0fcd" => :x86_64_linux
-  end
-
```
Comment (sjackman) [inline Formula/systemd.rb]: No need to remove the `bottle` section. Bump the source `revision` if necessary.

File: Formula/libusb.rb [hunk 4]
```diff
@@ -11,7 +11,6 @@ class Libusb < Formula
     sha256 "e1a0f90bb8906e3d833b033a4bed058f6aa1700d376971db700c8741527dafa9" => :el_capitan
     sha256 "6a4fb2012bf9106fcf1c71b215b711188d7c88d0d12b2cf744ba966363b2144d" => :yosemite
     sha256 "5a475e2ca93886e51b994d1ea323e915c91d8463e5b23b45203acb69edf69981" => :mavericks
-    sha256 "61de2c4a81871b5d0e7cb4fae460e93e2d407defbb41507bb9316cdfb9d3b6dd" => :x86_64_linux
```
Comment (sjackman) [inline Formula/libusb.rb]: No need to remove the bottle `sha256`. Bump the formula `revision` if necessary.

File: Formula/libusb.rb (hunk 1 above)
Comment (maxim-belkin) [inline Formula/libusb.rb]: Aren't `if OS.linux?` and `unless OS.mac?` equivalent?

File: Formula/systemd.rb (hunk 2 above)
Comment (maxim-belkin) [inline Formula/systemd.rb]: I don't think xml::parser module is installed on Macs.

**Comment by maxim-belkin** [comment]
Error on the CircleCI:
```
./libusb-1.0.20.x86_64_linux.bottle.2.tar.gz
  bottle do
    cellar :any
    revision 2
    sha256 "6a9805e3b8470e37e56173520d8d32f767bc084ee28dbcfda9c2eb6700331f63" => :x86_64_linux
  end
==> brew bottle --merge --write --no-commit ./libusb-1.0.20.x86_64_linux.bottle.json --keep-old
Error: --keep-old was passed but there were changes in sha256 => x86_64_linux!
==> libusb
==> FAILED
```
Two questions:
1. What is failing? `brew bottle --merge...`?
2. Why is `--keep-old` specified after the `json`?

File: Formula/libusb.rb (hunk 1 above)
Comment (sjackman) [inline Formula/libusb.rb]: Yep. Consider what the behaviour should be on a mythical third OS, like BSD or Windows Subsystem for Linux, when choosing between the two.

**Comment by maxim-belkin** [comment]
rebased the branch to see if latest changes fix the CircleCI build

File: Formula/systemd.rb (hunk 2 above)
Comment (rwhogg) [inline Formula/systemd.rb]: I thought it was: https://github.com/Linuxbrew/brew/issues/6#issue-152571915

Comment (rwhogg) [inline Formula/systemd.rb]: Either way, there's a *lot* of formulae that fail without XML::Parser. (For a long time, I couldn't use CPAN modules on my machine at all, which stopped me from building a ridiculous number of formulae). 

It'd be really tedious and repetitive to have to do this for all of them.

**Comment by sjackman** [comment]
> 1. What is failing? brew bottle --merge...?

Because there is already a bottle for `:x86_64_linux` and `brew bottle --keep-old` says to keep the existing bottles.

> 2. Why is --keep-old specified after the json?

Because usually we're adding an `:x86_64_linux` that already has bottles for Mac without changing the revision. When we do change the revision, things go sour.

File: Formula/libusb.rb [hunk 5]
```diff
@@ -7,7 +7,7 @@ class Libusb < Formula
 
   bottle do
     cellar :any
-    revision 1
+    revision 2
```
Comment (sjackman) [inline Formula/libusb.rb]: Don't change the bottle revision. Instead change the formula revision.

File: Formula/libusb.rb [hunk 6]
```diff
@@ -29,6 +29,7 @@ class Libusb < Formula
   deprecated_option "no-runtime-logging" => "without-runtime-logging"
 
   depends_on "systemd" if OS.linux? # for libudev
+  depends_on "zlib" if OS.linux?
```
Comment (sjackman) [inline Formula/libusb.rb]: Use instead `depends_on "zlib" unless OS.mac?`

File: Formula/systemd.rb [hunk 7]
```diff
@@ -1,11 +1,12 @@
 class Systemd < Formula
   desc "System and service manager"
   homepage "https://wiki.freedesktop.org/www/Software/systemd/"
-  url "http://www.freedesktop.org/software/systemd/systemd-221.tar.xz"
+  url "https://www.freedesktop.org/software/systemd/systemd-221.tar.xz"
   sha256 "085e088650afbfc688ccb13459aedb1fbc7c8810358605b076301f472d51cc4f"
   # tag "linuxbrew"
 
   bottle do
+    revision 2
```
Comment (sjackman) [inline Formula/systemd.rb]: Don't change the bottle revision. Instead change the formula revision.

File: Formula/systemd.rb [hunk 8]
```diff
@@ -18,8 +19,22 @@ class Systemd < Formula
   depends_on "util-linux" # for libmount
   depends_on "coreutils" => :build
   depends_on "XML::Parser" => :perl
+  depends_on "expat" => :build unless OS.mac?
```
Comment (sjackman) [inline Formula/systemd.rb]: Is it only a `:build` dependency? `libexpat.so` isn't needed at run time?

File: Formula/systemd.rb (hunk 2 above)
Comment (sjackman) [inline Formula/systemd.rb]: If there are bottles for Mac, then `XML::Parser` ships with Mac (or the bottles wouldn't build).

File: Formula/systemd.rb [hunk 9]
```diff
@@ -18,8 +19,22 @@ class Systemd < Formula
...
 
   def install
+    if ENV["TRAVIS"]
+      resource("xml::parser").stage do
+        system "perl", "Makefile.PL", "PREFIX=/home/linuxbrew/perl5"
+        system "make"
+        system "make", "install"
+      end
+    end
+
```
Comment (sjackman) [inline Formula/systemd.rb]: Please delete this chunk.

File: Formula/systemd.rb [hunk 10]
```diff
@@ -1,11 +1,12 @@
 class Systemd < Formula
   desc "System and service manager"
   homepage "https://wiki.freedesktop.org/www/Software/systemd/"
-  url "http://www.freedesktop.org/software/systemd/systemd-221.tar.xz"
+  url "https://www.freedesktop.org/software/systemd/systemd-221.tar.xz"
```
Comment (sjackman) [inline Formula/systemd.rb]: This change should go upstream to Homebrew.

File: Formula/systemd.rb (hunk 9 above)
Comment (maxim-belkin) [inline Formula/systemd.rb]: This chunk is the whole point of this PR. We might wanna hack the "depends_on" function to check for specific text that on Travic / CircleCI will do exactly that.

File: Formula/systemd.rb (hunk 10 above)
Comment (maxim-belkin) [inline Formula/systemd.rb]: I don't see it in Homebrew :/

Comment (sjackman) [inline Formula/systemd.rb]: Sorry, I was confusing `libusb` with `systemd`. You're right. `systemd` is only in Linuxbrew.

File: Formula/systemd.rb (hunk 8 above)
Comment (maxim-belkin) [inline Formula/systemd.rb]: I see two `.so` files: `Cellar/systemd/221/lib/libsystemd.so` and `Cellar/systemd/221/lib/libudev.so` and none of them depends on libexpat in my installation.

Comment (sjackman) [inline Formula/systemd.rb]: I've checked and I can't find any run-time dependency on expat either. Are you sure that it's a dependency at all? What error do you get without the dependency on `expat`?

Comment (maxim-belkin) [inline Formula/systemd.rb]: On my machine `expat` is necessary in order for the `XML::Parser` dependency to be detected (weird). Without it I get:
```
systemd: Unsatisfied dependency: XML::Parser
Homebrew does not provide special Perl dependencies; install with:
  `cpan -i XML::Parser`
```

**Comment by maxim-belkin** [comment]
rebased the PR against the latest master branch of the Linuxbrew

**Pull request closed without merging**